### PR TITLE
[Speedy] do not need disclosed contract at Machine creation time

### PR DESCRIPTION
**Pull request by remyhaemmerle-da**
This PR makes several changes on how the disclosed contract are handle by the speedy machine.

(1) disclosed contract at not anymore pass a builtin time. They are anyway passed latter using `SBCacheDisclosedContract`.

(2) disclosed contracts are first keep in their own cache, `disclosedContracts: Map[V.ContractId, CachedContract]`. They are moved to the cached for used contract only if used.  In fact we need to be sure the disclosed contracts pass through the `Speedy#addGlobalContract` to verify the transaction limits are not exceeded iff they are used. 

(3) duplicate key for disclosed contracts is check during preprocessing.

### CODE REVIEW

File: daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/Engine.scala [hunk 1]
```diff
@@ -372,8 +373,17 @@ class Engine(val config: EngineConfig = Engine.StableConfig) {
...
             ) =>
+          val disclosureMap = disclosures.iterator.map(c => c.contractId.value -> c).toMap
+          val processedDisclosedContracts = disclosedCreateEvents.map { create =>
+            val diclosedContract = disclosureMap(create.coid)
+            ProcessedDisclosedContract(
+              create,
+              diclosedContract.metadata.createdAt,
+              diclosedContract.metadata.driverMetadata,
+            )
+          }
```
Comment (remyhaemmerle-da) [inline daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/Engine.scala:386]: The engine is now responsible to add the meta data to the used disclosed contracts. There is no reason to pass those info to Speedy.

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala [hunk 2]
```diff
@@ -1149,33 +1149,40 @@ private[lf] object SBuiltin {
...
-                  // of actualTmplId if not known.
-                  SEVal(ToCachedContractDefRef(actualTmplId)),
-                  SELet1(
-                    SEImportValue(Ast.TTyCon(actualTmplId), arg),
-                    SEAppAtomic(SELocS(2), Array(SELocS(1), SEValue(args.get(1)))),
-                  ),
+          machine.disclosedContracts.get(coid) match {
+            case Some(contract) =>
+              machine.addGlobalContract(coid, contract)
+              Control.Value(contract.any)
```
Comment (remyhaemmerle-da) [inline daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala:1155]: Here the disclosed contract are adding to the cached contracts while transaction limits are enforced

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Speedy.scala [hunk 3]
```diff
@@ -472,47 +468,10 @@ private[lf] object Speedy {
...
-
-    private[speedy] def addContractKey(
-        templateId: TypeConName,
-        keyHash: crypto.Hash,
-        contractId: V.ContractId,
-    ): Either[IError, Unit] = {
-      if (keyMap.contains(keyHash)) {
-        Left(
-          IError.DisclosurePreprocessing(
-            IError.DisclosurePreprocessing.DuplicateContractKeys(
```
Comment (remyhaemmerle-da) [inline daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Speedy.scala:491]: No reason to checks that at runtime. This should be done at preprocessing time.

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala [hunk 4]
```diff
@@ -2127,25 +2132,10 @@ private[lf] object SBuiltin {
     override protected def executeUpdate(
         args: util.ArrayList[SValue],
         machine: UpdateMachine,
-    ): Control[Nothing] = {
+    ): Control.Value = {
```
Comment (remyhaemmerle-da) [inline daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala:2135]: Here the whole contract is pass as argument to the builtin. Then we do not need the disclosure table to be prefilled.

File: daml-lf/spec/daml-lf-1.rst [hunk 5]
```diff
@@ -718,13 +718,13 @@ Then we can define our kinds, types, and expressions::
   Updates
     u ::= 'pure' @τ e                               -- UpdatePure
        |  'bind' x₁ : τ₁ ← e₁ 'in' e₂               -- UpdateBlock
-       |  'create' @Mod:T e                         -- UpdateCreate
+  x     |  'create' @Mod:T e                         -- UpdateCreate
```
Comment (remyhaemmerle-da) [inline daml-lf/spec/daml-lf-1.rst:721]:
```suggestion
       |  'create' @Mod:T e                         -- UpdateCreate
```

**Review by carlpulley-da — APPROVED** [review]

File: daml-lf/spec/daml-lf-1.rst [hunk 6]
```diff
@@ -718,13 +718,13 @@ Then we can define our kinds, types, and expressions::
...
        |  'create_interface' @Mod:I e               -- UpdateCreateInterface [Daml-LF ≥ 1.dev]
        |  'fetch' @Mod:T e                          -- UpdateFetch
        |  'fetch_interface' @Mod:I e                -- UpdateFetchInterface [Daml-LF ≥ 1.dev]
-       |  'exercise' @Mod:T Ch e₁ e₂                -- UpdateExercise
-       |  'exercise_by_key' @Mod:T Ch e₁ e₂         -- UpdateExerciseByKey [Daml-LF ≥ 1.11]
-       |  'exercise_interface' @Mod:I Ch e₁ e₂ e₃   -- UpdateExerciseInterface [Daml-LF ≥ 1.dev]
+  x     |  'exercise' @Mod:T Ch e₁ e₂                -- UpdateExercise
+  x     |  'exercise_by_key' @Mod:T Ch e₁ e₂         -- UpdateExerciseByKey [Daml-LF ≥ 1.11]
+  x     |  'exercise_interface' @Mod:I Ch e₁ e₂ e₃   -- UpdateExerciseInterface [Daml-LF ≥ 1.dev]
```
Comment (remyhaemmerle-da) [inline daml-lf/spec/daml-lf-1.rst:727]:
```suggestion
       |  'exercise' @Mod:T Ch e₁ e₂                -- UpdateExercise
       |  'exercise_by_key' @Mod:T Ch e₁ e₂         -- UpdateExerciseByKey [Daml-LF ≥ 1.11]
       |  'exercise_interface' @Mod:I Ch e₁ e₂ e₃   -- UpdateExerciseInterface [Daml-LF ≥ 1.dev]
```

[the remark this answers is not in the document]
Comment (remyhaemmerle-da) [reply daml-lf/spec/daml-lf-1.rst]: This have been committed by error. I reverted it. 
Thanks.

**Pull request merged**